kubeadm: imagePullPolicy option in init config

rosti opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR adds `imagePullPolicy` option to the `kubeadm init` configuration file.

The new `imagePullPolicy` option is forwarded to the generated kubelet static pods for etcd, kube-apiserver, kube-controller-manager and kube-scheduler. This option allows for precise image pull policy specification for master nodes and thus for more tight control over images. It is useful in CI environments and in environments, where the user has total control over master VM templates (thus, the master VM templates can be preloaded with the required Docker images for the control plane services).

**Special notes for your reviewer**:

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/assign @luxas

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
kubeadm: New "imagePullPolicy" option in the init configuration file, that gets forwarded to kubelet static pods to control pull policy for etcd and control plane images.
```

rosti committed:
kubeadm: Support imagePullPolicy option in the kubeadm init configuration file

rosti commented:
/assign @timothysc

jamiehannaford commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -63,6 +64,9 @@ type MasterConfiguration struct {
 	// ImageRepository what container registry to pull control plane images from
 	ImageRepository string
 
+	// ImagePullPolicy that control plane images. Can be Always, IfNotPresent or Never.
```
for control plane images

jamiehannaford commented on cmd/kubeadm/app/apis/kubeadm/v1alpha1/types.go:
```diff
@@ -62,6 +63,8 @@ type MasterConfiguration struct {
 	// ImageRepository what container registry to pull control plane images from
 	ImageRepository string `json:"imageRepository"`
+	// ImagePullPolicy that control plane images. Can be Always, IfNotPresent or Never.
+	ImagePullPolicy v1.PullPolicy `json:"imagePullPolicy,omitempty" protobuf:"bytes,14,opt,name=imagePullPolicy,casttype=v1.PullPolicy"`
```
we don't need the protobuf tag

jamiehannaford reviewed (changes requested):
What happens if someone sets ImagePullPolicy=Never on a node without the images? The control plane pods will be stuck in `ErrImageNeverPull` right?

If we want to allow Never, we should allow users to find out what images _will_ be pulled, and then verify they exist locally first.

errordeveloper commented:
@rosti thanks, this is useful. I agree with @jamiehannaford, as we still don't have an image lister command of any sort, have you considered this?

timothysc reviewed (changes requested):
If we are going to do this we should be explicit about the defaults, also address other comments.

rosti commented:
Rebased on latest master and made the changes that @jamiehannaford suggested.

I checked the behavior with imagePullPolicy set to Never and no images present. Kubeadm times out (after ~30 minutes) with the very generic error message `kubeletFailTempl` (defined in `cmd/kubeadm/app/cmd/init.go`). That message basically says "check with the kubelet service and logs to find what's wrong".
The kubelet logs, then, contain an error messages in the form `Container image \"xyz:version\" is not present with pull policy of Never`. This may be sufficient for now, but having error messages, that are immediate (as opposed to after some timeout) and more detailed on what went wrong is very nice.

Do you think, that we need to do the more comprehensive checks now or, possibly, leave this for another PR?

Having no value specified in the init config file for imagePullPolicy, leaves the exact decision to kubelet for this. It is the way things are done currently.

timothysc commented:
@rosti - Just so I'm clear, if ImagePullPolicy is not specified it will default to nil and works as it is today?

timothysc reviewed (changes requested):
please squash commits and verify minor question then lgtm

rosti commented:
Rebased on latest master and squashed commits (as per request).

@timothysc, you are right. Not setting imagePullPolicy in init config, will default it to nil and thus it won't serialize it to static pod YAMLs. Then kubelet will work just as it does now.
So not setting this option in init config won't change the current behavior.

jamiehannaford commented:
I'm still not comfortable with setting up users to fail with really unhelpful errors. How about we generate a warning if the user provides imagePullPolicy=Never? It could just list the image names listed [here](https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/images/images.go#L38-L43). That would be far better UX IMO and not too hard to implement. WDYT @rosti?

rosti commented:
@jamiehannaford, looks nice, will do that tomorrow.
I was also thinking about reducing the timeout for services to come up when imagePullPolicy == Never. Currently it's ~30 minutes, but that time includes possible download over the Internet (which we don't have when imagePullPolicy is Never). A couple of minutes should be enough.
What do you think?

rosti commented:
@jamiehannaford, @timothysc, can you take a look at this?

I expanded the [kubeletFailTempl](https://github.com/rosti/kubernetes/blob/a2bbc27c0c931b81ea973906c48d6b590bfbaedd/cmd/kubeadm/app/cmd/init.go#L95) error message, to mention the possibility of missing images if imagePullPolicy is set to Never and to contain the image of etcd.
Also, reduced the timeout of the waiter, returned by [getWaiter](https://github.com/rosti/kubernetes/blob/d0e84fc3cd65c6792544a5afae0bd94ae206c958/cmd/kubeadm/app/cmd/init.go#L529), from 30 minutes to 60 seconds if imagePullPolicy is Never.

And, of course, rebased on latest master.

timothysc reviewed (approved):
I'll defer to @jamiehannaford on lgtm here, I don't have strong feelings b/c I think this will be an exception to the norm.  

However please squash your commits. 

/approve

jamiehannaford reviewed (approved):
/lgtm

rosti commented:
Fixed build due to missing dependencies.

xiangpengzhao commented:
@rosti PR needs rebase :)

rosti commented:
/retest

fejta commented:
@timothysc removing the needs-ok-to-test label does not make a PR ok-to-test.

Please use commands to mutate labels.

jamiehannaford commented:
/ok-to-test
